### ISSUE TITLE
evhtp_send_reply_chunk evhtp_send_reply   evhtp_send_reply_chunk_end memoryleak

### ISSUE THREAD

**Issue by sunchengbi** [criticalstack/libevhtp#47]
# Details
 memory leak
# Steps or code to reproduce the problem.
code as flows:

```C
 void
      reqHandler(evhtp_request_t * req, void * arg)
      {
          struct evbuffer * buffer    = evbuffer_new();
          char              buf[4096] = { 0 };
          int               i;
      
          evhtp_send_reply_chunk_start(req, EVHTP_RES_OK);
      
          for (i = 0; i < 10000; i++)
          {
              evbuffer_add(buffer, buf, sizeof(buf));
              evhtp_send_reply_chunk(req, buffer);
              evbuffer_drain(buffer, -1);
          }
      
          evhtp_send_reply_chunk_end(req);
      
          evbuffer_free(buffer);
      }
      
      int main(int argc,char **argv)
      {
              evbase_t          * evbase;
              evhtp_t           * evhtp;
      
              evbase            = event_base_new();
              assert(evbase);
              evhtp             = evhtp_new(evbase, NULL);
              assert(evhtp);
      
              evhtp_set_gencb(evhtp, reqHandler, NULL);
              int ret = evhtp_use_threads(evhtp, NULL, 64, NULL);
      
              ret = evhtp_bind_socket(evhtp, "0.0.0.0", 9010, 1024);
      
              ret = event_base_loop(evbase, 0);
      }
```

# Version
 1.2.12

**Comment by sftwrngnr** [comment]
Looking into it. I will get back to you.

**Comment by sunchengbi** [comment]
I also found antoher problem that memory is very large before executing this line of code. 

evhtp_send_reply_chunk_end(req);

**Comment by wolfchimneyrock** [comment]
I was having a similar issue, but after looking at https://github.com/erikarn/libevhtp-http/blob/master/src/clt/clt.c i realize the proper usage is to set
 
evbuffer_set_flags(buffer, EVBUFFER_FLAG_DRAINS_TO_FD);

on this intermediate buffer before doing any evbuffer_add_file_segment operations, otherwise libevent will copy the data out instead of using sendfile or memmap.

The libevent documentation states that you should never have to set this manually, yet not setting it causes huge memory overhead when doing evbuffer_add_file_segment(..)

**Comment by NathanFrench** [comment]
Thank you @wolfchimneyrock!

**Comment by ousado** [comment]
There indeed seems to be strange behavior here regarding memory usage, but it's not an actual leak. 
Running this example with different workloads, sometimes memory usage after a given sequence of requests stays relatively high. That usage somewhat correlates with the number of concurrent requests. In one example session (several runs of requests, **without** restarting the server), I got these numbers, for instance:
100 requests with 30 concurrent connections: 629M RES
100 requests with  2 concurrent connections: 3.5K RES
100 requests with 30 concurrent connections: 3.5K RES
150 requests with 30 concurrent connections: 3.5K RES
150 requests with 30 concurrent connections: 2195M RES
150 requests with 30 concurrent connections: 3.5K RES

It's not a memory leak, because the memory definitely is still reachable and is recycled eventually, however an inactive server could potentially hold on to much more memory than required. Repeating the above experiment while running another client issuing a single request each second in parallel shows that memory usage always returns to ~3.5K on every such "standalone" request. 

Interestingly and strangely this behavior doesn't seem to occur at all when a `evhtp_hook_on_connection_fini` hook is registered. 


@sunchengbi 
You seem to assume that you're dealing with a synchronous networking API here - that's not the case. When you're using `evhtp_send_reply_chunk` the data is queued for sending, not necessarily sent immediately: When you're queuing 40MB per request (like in this example - or almost 2GB like in one of the examples in another issue you opened here) at once, you'll obviously have a lot of allocated memory until the data is sent and the respective buffers are freed. 
What you're doing amounts to a misuse of the API. What you should do instead is register an evhtp_hook_on_chunk_complete hook (and an evhtp_hook_on_chunks_complete hook to do the final cleanup) to send each subsequent chunk after the preceding one has been written out.

**Comment by NathanFrench** [comment]
Closing, not a bug: Please see @ousado epic reply.

Thanks!

**Comment by NathanFrench** [comment]
I added a quick example (with lots of comments) on how to stream large amounts of data without consuming memory using chunked replies here: https://github.com/criticalstack/libevhtp/blob/develop/examples/example_chunked.c

Cheers!
